core/rawdb/chain_freezer.go

```diff
@@ -77,6 +73,54 @@ func (f *chainFreezer) Close() error {
 	return f.Freezer.Close()
 }
 
+// block tags
+const (
+	finalizedBlock = "finalized"
+	headBlock      = "head"
+)
+
+// readBlock returns the number of specified block. 0 is returned if the block
+// is unknown or not available yet.
+func (f *chainFreezer) readBlock(db ethdb.KeyValueReader, tag string) uint64 {
+	var hash common.Hash
+	switch tag {
+	case finalizedBlock:
+		hash = ReadFinalizedBlockHash(db)
+	case headBlock:
+		hash = ReadHeadBlockHash(db)
+	default:
+		return 0
+	}
+	if hash == (common.Hash{}) {
+		return 0
```

Comment:
This might warrant a log.Error, at least in the case of head block, since that should always exit?